crates/sc-proof-of-time/src/source.rs

```diff
@@ -35,18 +41,44 @@ pub struct PotSourceConfig {
 /// Depending on configuration may produce proofs of time locally, send/receive via gossip and keep
 /// up to day with blockchain reorgs.
 #[derive(Debug)]
-pub struct PotSource {
-    // TODO
+pub struct PotSource<Block, Client> {
+    // TODO: Use this in `fn run`
+    #[allow(dead_code)]
+    client: Arc<Client>,
+    _block: PhantomData<Block>,
 }
 
-impl PotSource {
-    pub fn new(config: PotSourceConfig) -> (Self, PotSlotInfoStream) {
+impl<Block, Client> PotSource<Block, Client>
+where
+    Block: BlockT,
+    BlockHash: From<Block::Hash>,
+    Client: ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+    Client::Api: SubspaceRuntimeApi<Block, FarmerPublicKey>,
+{
+    pub fn new(
+        config: PotSourceConfig,
+        client: Arc<Client>,
+    ) -> Result<(Self, PotSlotInfoStream), ApiError> {
+        let PotSourceConfig {
+            // TODO: Respect this boolean flag
+            is_timekeeper: _,
+            initial_key,
+        } = config;
+        let info = client.info();
         // TODO: All 3 are incorrect and should be able to continue after node restart
         let start_slot = SlotNumber::MIN;
-        let start_seed = PotSeed::default();
-        let start_key = config.initial_key;
-        // TODO: Change to correct values taken from blockchain
-        let iterations = NonZeroU32::new(1024).expect("Not zero; qed");
+        let start_seed = PotSeed::from_genesis_block_hash(BlockHash::from(info.genesis_hash));
+        let start_key = initial_key;
+        #[cfg(feature = "pot")]
+        let best_hash = info.best_hash;
+        #[cfg(feature = "pot")]
+        let runtime_api = client.runtime_api();
+        #[cfg(feature = "pot")]
+        let iterations = runtime_api
+            .pot_parameters(best_hash)?
+            .iterations(Slot::from(start_slot));
+        #[cfg(not(feature = "pot"))]
```

Comment:
Not sure I understand the conditional compilation here. Why just make `PotSource` under `pot` instead since `new` is also under `pot` flag 🤔

Reply:
The idea here is that the whole `sc-proof-of-time` should generally be functional by default since we do compile it with the rest fo the codebase. The challenge though is that it uses some conditionally compiled APIs like this runtime API call, hence sometimes we put some placeholder values here.

However more generally we are still able to use it, for example, to run tests, without requiring `pot` feature.

You could say tests can also be enabled just for `pot` feature (the opposite is already done for some crates until they are compatible), but I decided to go with this approach instead.